### DISCUSSION TITLE
Bound Checks slowing code down?

### DISCUSSION THREAD

**Discussion by matzhaugen** [rust-ndarray/ndarray#868] [Q&A]
Hey I was wondering, do the bound checks for all assignment operations slow down the code? For example if I did
```
use ndarray::Array2;
let p: usize = 100;
let matrix1 = Array2::<f64>::zeros((p, p));
let mut matrix2 = Array2::<f64>::ones((p,  p));

matrix2 = &matrix1;
```
would I do `p*p` bound checks?

If so how can one avoid this without doing `unsafe` rust?

**Comment by bluss** [comment]
Since it's a question, I moved it over here to the Q&A section. It helps keeps the issue list focused.

Your code doesn't compile :) Can you fix that and mark which line you are talking about? Variable assignments/value moves in Rust don't run any library defined code at all, so that doesn't bounds check.

**Reply by matzhaugen** [reply]
Pardon me, here is code that compiles.
```
use ndarray::Array2;
use std::time::Instant;
fn main() {
    let p: usize = 100;
    let matrix1 = Array2::<f64>::zeros((p, p));
    let mut matrix2 = Array2::<f64>::ones((p, p));
    let start = Instant::now();
    matrix2 = -(&matrix1);
    let duration = start.elapsed();
    println!("Time elapsed is: {:?}", duration);
}
```
This runs in .5 ms, which I would say is quite slow, and thus leading me to think that there is bound checking.

**Reply by mulimoen** [reply]
The problem lies here:
```rust
let mut matrix2 = Array2::<f64>::ones((p, p)); // Result is not used
matrix2 = -(&matrix1); // a new matrix
```
`matrix2` in the first binding is never used, as it is overwritten by the second assignment. You probably want to use
```rust
let mat1 = Array2::<i32>::zeros((2, 2));
let mut mat2 = mat1.clone();
    
azip!((m1 in &mat1, m2 in &mut mat2) *m2 = -m1);
```
which uses the storage in `mat2`  efficiently. This should not give more than one bounds check and uses efficient iterators.

**Reply by matzhaugen** [reply]
Hey @mulimoen! Unfortunately that slowed the code down. 
I made a little repo to illustrate my example https://github.com/matzhaugen/ndarray-bench . If you have a mac and g++ installed it should be easy to run.

**Reply by mulimoen** [reply]
I ser you use `cargo run` to run the binary. This will run a non-optimized binary. Either run `target/release/foo` or use `cargo run --release`

**Reply by matzhaugen** [reply]
That increased the speed by 500x. My bad, thanks for clarifying :)

**Reply by matzhaugen** [reply]
Why isn't this zip contraption the default in the assignment operations? It seems like it should be the norm. Otherwise, one has to write this zip expression for all assignments if you want the code to be fast, no?

**Reply by bluss** [reply]
There is no evidence to suggest that the zip thing is required to make the unary negation operation fast.

Also, again, you can't "overload" assignment in Rust, and assignment is not where the interesting operations happen here. For example the unary minus operation is "interesting".
